### PR TITLE
[Core][ModelPart] const version of pNodes

### PR DESCRIPTION
**Pull request by philbucher**
part of #2181

### CODE REVIEW

**Comment by philbucher** [comment]
@KratosMultiphysics/technical-committee maybe we should also change this in the python interface to use the `const` method

Or maybe a deprecation warning there first?

=>https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python/add_model_part_to_python.cpp#L119-L127

**Comment by philbucher** [comment]
its not compiling bcs `Mesh` has a property `Nodes`:
`.def_property("Nodes", &MeshType::pNodes,&MeshType::SetNodes)`

Should this point to the const or non-const version?

**Comment by philbucher** [comment]
I commented the pNodes functions in the MP to see where they are used and it seems like there are not too many places (besides the py-interface):
- StokesIntiProcess (FluidDynamicsApp & TrilinosApp)
- TrilinosSpalartAlmarasTurbModel (TrilinosApp)

I compiled the following Apps:
~~~
STRUCTURAL_MECHANICS_APPLICATION
FLUID_DYNAMICS_APPLICATION
MESH_MOVING_APPLICATION
FSI_APPLICATION
EMPIRE_APPLICATION
DEM_APPLICATION
PARTICLE_MECHANICS_APPLICATION
EXTERNAL_SOLVERS_APPLICATION
MESHING_APPLICATION
SHAPE_OPTIMIZATION_APPLICATION
MAPPING_APPLICATION
EIGEN_SOLVERS_APPLICATION
HDF5_APPLICATION
TRILINOS_APPLICATION
METIS_APPLICATION
~~~

**Comment by RiccardoRossi** [comment]
> its not compiling bcs `Mesh` has a property `Nodes`:
> `.def_property("Nodes", &MeshType::pNodes,&MeshType::SetNodes)`
> 
> Should this point to the const or non-const version?

i would say that we should use the const version here as in the other places. another thing is if this will work...

**Comment by philbucher** [comment]
Assuming that we can still modify the Nodes if we return the const version I thought of the following:
For ModelPart we should return the const version to prevent e.g. push_back which breaks the hierarchy
For Mesh though I think that we should return the non-const version because there a modification of the container is allowed I think

We should discuss this 

BTW does the same that we are discussing here for Nodes also apply to Elements/Constraints/Conditions/Properties?

**Comment by RiccardoRossi** [comment]
@philbucher i think your point is fair. Mesh is a lower level entity, so it should allow lower level manipulation

File: kratos/includes/model_part.h [hunk 1]
```diff
@@ -475,7 +475,12 @@ class KRATOS_API(KRATOS_CORE) ModelPart : public DataValueContainer, public Flag
         return GetMesh(ThisIndex).Nodes();
     }
 
-    NodesContainerType::Pointer pNodes(IndexType ThisIndex = 0)
+    KRATOS_DEPRECATED_MESSAGE("This is legacy version") NodesContainerType::Pointer pNodes(IndexType ThisIndex = 0)
```
Comment (RiccardoRossi) [inline kratos/includes/model_part.h:495]: this is ok to me, nevertheless let's see if we have quorum in @KratosMultiphysics/technical-committee

**Review by RiccardoRossi — COMMENTED** [review]
regarding the python interface, what about throwing an error the same way as we do for the SubModelParts property?

or define it as readonly_property

**Comment by rubenzorrilla** [comment]
@philbucher CI crashes.

**Comment by philbucher** [comment]
> its not compiling bcs `Mesh` has a property `Nodes`:
> `.def_property("Nodes", &MeshType::pNodes,&MeshType::SetNodes)`
> 
> Should this point to the const or non-const version?

@rubenzorrilla ;)
there are things to be discussed/solved for it to compile

**Comment by roigcarlo** [comment]
@KratosMultiphysics/technical-committee agrees on deprecating the `pNodes` method as it bypasses the coherence of the submodelparts and their parents.

**Comment by philbucher** [comment]
ok so now I implemented the following:
> Assuming that we can still modify the Nodes if we return the const version I thought of the following:
> For ModelPart we should return the const version to prevent e.g. push_back which breaks the hierarchy
> For Mesh though I think that we should return the non-const version because there a modification of the container is allowed I think

BTW does the same that we are discussing here for Nodes also apply to Elements/Constraints/Conditions/Properties?

**Comment by philbucher** [comment]
ok I gave this another go and I suggest we merge this as is

The other problems I pointed out are only in the python-interface (i.e. the way the [PointerVectorSet is exposed](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python/containers_interface.h#L100-L123)) and will be addressed in a future PR

**Comment by philbucher** [comment]
@KratosMultiphysics/technical-committee decided that I will try to expose the `Nodes`, `Conditions` and `Elements` as read-only property instead of just property. (The `Constraints` are already exposed like this)
Then we see how it affects the code and evaluate

**Pull request closed without merging**